File: 0337. House Robber III/tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    static class RobInfo{
        public int curMax;
        public int notRobMax;
        public RobInfo(int _curMax, int _notRobMax){
            this.curMax = _curMax;
            this.notRobMax = _notRobMax;
        }
    }
    private RobInfo robNode(TreeNode root){
        if(root == null){
            return new RobInfo(0, 0);
        }
        RobInfo left = robNode(root.left);
        RobInfo right = robNode(root.right);
        int robMax = root.val + left.notRobMax + right.notRobMax;
        int notRobMax = left.curMax + right.curMax;
        int curMax = Math.max(robMax, notRobMax);
        return new RobInfo(curMax, notRobMax);
    }
    public int rob(TreeNode root) {
        RobInfo robInfo = robNode(root);
        return Math.max(robInfo.curMax, robInfo.notRobMax);
    }
}

File: 0337. House Robber III/tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int robNode(TreeNode* root, int& robMax, int& notRobMax){
        if(root == nullptr)
            return 0;
        int leftRobMax = 0, leftNotRobMax = 0, rightRobMax = 0, rightNotRobMax = 0;
        int leftMax = robNode(root->left, leftRobMax, leftNotRobMax);
        int rightMax = robNode(root->right, rightRobMax, rightNotRobMax);
        robMax = root->val + leftNotRobMax + rightNotRobMax;
        notRobMax = leftMax + rightMax;
        return max(robMax, notRobMax);
    }
    
    int rob(TreeNode* root) {
        int robMax = 0, notRobMax = 0;
        return robNode(root, robMax, notRobMax);
    }
};